### ISSUE TITLE
Generation of accept and responseType should check all type of responses in the YAML

### ISSUE THREAD

**Issue by forestmaxime** [cyclosproject/ng-openapi-gen#115]
First of all, thanks @luisfpg for all the work on this project, it helped us save a lot of time to bridge our UI and API.

I've come across an issue with the generation of the properties "accept" and "responseType". We use a subset of the spec for [JSON:API Errors](https://jsonapi.org/format/#errors) when there's an error in our API we will return a JSON.

**Solutions proposed:**
"accept" should be generated from a concatenation all responses
"responseType" should probably be empty for 200 OK (with no body) OR 204 No Content and let Angular HTTP Client handle the responseType.

**YAML**
```
paths:
  /api/v1/account:
    delete:
      tags:
        - account
      summary: ""
      operationId: delete_account
      requestBody:
        description: DeleteRequest
        required: true
        content:
          application/json:
            schema:
              $ref: "#/components/schemas/DeleteRequest"
      responses:
        "404":
          description: Not Found
          content:
            application/json:
              schema:
                $ref: "#/components/schemas/JsonApiError"
        "500":
          description: Internal Server Error
          content:
            application/json:
              schema:
                $ref: "#/components/schemas/JsonApiError"
        "400":
          description: Bad Request
          content:
            application/json:
              schema:
                $ref: "#/components/schemas/JsonApiError"
        "204":
          description: No Content
```

**When this gets generated in Angular:**
```
deleteAccount$Response(params: {
  
    /**
     * DeleteRequest
     */
    body: DeleteRequest
  }): Observable<StrictHttpResponse<void>> {

    const rb = new RequestBuilder(this.rootUrl, AccountService.DeleteAccountPath, 'delete');
    if (params) {


      rb.body(params.body, 'application/json');
    }
    return this.http.request(rb.build({
      responseType: 'text',
      accept: '*/*'
    })).pipe(
      filter((r: any) => r instanceof HttpResponse),
      map((r: HttpResponse<any>) => {
        return (r as HttpResponse<any>).clone({ body: undefined }) as StrictHttpResponse<void>;
      })
    );
  }
```

**Comment by forestmaxime** [comment]
My colleague and I came up with something that could most likely fix my issue:

https://github.com/cyclosproject/ng-openapi-gen/blob/master/lib/operation-variant.ts
```
const accept = [];
​
if (successResponse) {
	this.resultType = successResponse.type;
	this.responseType = this.inferResponseType(successResponse.mediaType);
	accept.push(successResponse.mediaType);
} else {
	this.resultType = null; // Since there's no content, we let Angular handle it
}
​
// for each errorResponse in errorResponses -> accept.push(this.inferResponseType(errorResponse.mediaType));
​
this.accept = accept.length > 0 ? accept.join(', ') : '*/*';
```

**Comment by luisfpg** [comment]
I don't see how that can help, because you do have a success response you your example (201).
Also, see this excerpt from the [Angular documentation](https://angular.io/api/common/http/HttpClient):

> The responseType value determines how a successful response body is parsed.

So, the responseType only affects success responses, not error responses.

**Comment by Samuel-B-D** [comment]
@luisfpg The point here is that the current behavior of ng-openapi-gen is to parse the response body according to the success responseType independantly of the errors. It means that if you have an endpoint with a success JSON body and an error text, you will try to deserialize the text as JSON and it will fail. Inversely, if your endpoint have a success type of 201 without a body, but a JSON error, since ng-openapi-gen treat an empty success body as responseType: text, it will NOT deserialize the JSON error body.

Hope that is clearer when worded that way.

**Comment by luisfpg** [comment]
The point is that the responseType's default is not 'auto detect', but rather 'json'.
So, if you do the request with no responseType (which will default to json) and you server returns `Content-Type: text/plain`, you will get a client-side error anyway, because Angular tries to parse that text as JSON, and will fail (of course, assuming the returned text isn't a valid JSON).

And that's how Angular's HTTP client works: if you set responseType: 'json' (the default) Angular will also try to parse the errors as JSON. If you set responseType: 'text', so you can handle success responses as text, any errors will also be passed in as text.

In our usages, we always have errors as JSON, but the success response varies. So, we handle it like this: https://github.com/cyclosproject/cyclos4-ui/blob/65ee1600b624de09fe093d25c44b9c3a6b09522c/src/app/core/error-handler.service.ts#L67 because the HttpErrorResponse.error can be a parsed JSON or a string.

So, to summarize: not passing in a responseType is the same as responseType: 'json', which will throw a client-side error for trying to parse the response as JSON if it doesn't conform. As such, we have no other choice but to keep the current implementation as is.
